File: go/list-ops/list_ops.go
// Package listops provides methods for operating on the IntList type
package listops

import "slices"

// IntList is an abstraction of a list of integers which we can define methods on
type IntList []int

// Foldl feeds each of its IntList elements, from left to right, into the given closure
// building an accumulated value, returning that value when all elements have been processed
func (s IntList) Foldl(fn func(int, int) int, initial int) int {
	if len(s) == 0 { // early return for an empty list
		return initial
	}
	reduction := initial

	for _, val := range s {
		reduction = fn(reduction, val)
	}

	return reduction
}

// Foldr feeds each of its IntList elements, from right to left, into the given closure
// building an accumulated value, returning that value when all elements have been processed
func (s IntList) Foldr(fn func(int, int) int, initial int) int {
	if len(s) == 0 { // early return for an empty list
		return initial
	}
	reduction := initial
	length := len(s)

	for idx := range s {
		reduction = fn(s[length-1-idx], reduction)
	}

	return reduction
}

// Filter accepts a closure that returns true or false for
// each element in the IntList. Only elements yielding true
// will be returned in the new IntList
func (s IntList) Filter(fn func(int) bool) IntList {
	if len(s) == 0 { // early return for an empty list
		return s
	}
	var buffer IntList
	slices.Grow(buffer, len(s))
	for _, val := range s {
		if fn(val) {
			buffer = append(buffer, val)
		}
	}

	return buffer
}

// Length returns the length of its IntList
func (s IntList) Length() int {
	return len(s)
}

// Map returns a new slice populated with the returned values
// from the given closure. If the IntList is empty, it will return
// that IntList
func (s IntList) Map(fn func(int) int) IntList {
	if len(s) == 0 { // early return for an empty list
		return s
	}
	length := len(s)
	buffer := make(IntList, length, length)
	for idx, val := range s {
		buffer[idx] = fn(val)
	}
	return buffer
}

// Reverse returns its IntList in reverse order
func (s IntList) Reverse() IntList {
	length := len(s)
	buffer := make(IntList, length, length)
	for idx, val := range s {
		buffer[length-1-idx] = val
	}
	return buffer
}

// Append appends the contents of lst to its IntList
func (s IntList) Append(lst IntList) IntList {
	return append(s, lst...)
}

// Concat combines the lists param with its
// IntList returning one flattened list
func (s IntList) Concat(lists []IntList) IntList {
	buffer := s
	slices.Grow(buffer, len(lists))
	for _, val1 := range lists {
		buffer = append(buffer, val1...)
	}
	return buffer
}

File: erlang/list-ops/src/list_ops.erl
-module(list_ops).

-compile({no_auto_import, [length/1]}).

-export([
    append/2,
    concat/1,
    filter/2,
    length/1,
    map/2,
    foldl/3,
    foldr/3,
    reverse/1
]).

append(T1, T2) ->
    case {T1, T2} of
        {[], Val} ->
            Val;
        {Val, []} ->
            Val;
        {Val1, Val2} ->
            [H | Tail1] = Val1,
            [H | append(Tail1, Val2)]
    end.

concat(List) ->
    case List of
        [] -> [];
        [H | T] -> append(H, concat(T))
    end.

% Keeping these two seperate for readability
filter(_, []) ->
    [];
filter(Function, [H | T]) ->
    case Function(H) of
        true -> [H | filter(Function, T)];
        false -> filter(Function, T)
    end.

length(List) ->
    case List of
        [] -> 0;
        [_ | T] -> 1 + length(T)
    end.

map(Function, List) ->
    case List of
        [] -> [];
        [H | T] -> [Function(H) | map(Function, T)]
    end.

foldl(Function, Start, List) ->
    case List of
        [] -> Start;
        [H | T] -> foldl(Function, Function(H, Start), T)
    end.

foldr(Function, Start, List) ->
    case List of
        [] -> Start;
        _ -> foldl(Function, Start, reverse(List))
    end.

reverse(List) ->
    case List of
        [] -> [];
        _ -> reverse(List, [])
    end.

reverse(List, Acc) ->
    case List of
        [] -> Acc;
        [H | T] -> reverse(T, [H | Acc])
    end.
